src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -807,253 +844,383 @@ RouterSelectTask(Query *query)
+		{
+			break;
+		}
+
+		relationRestriction = NULL;
 	}
 
-	/* make sure partition column is used only once in the query */
-	varClauseList = pull_var_clause_default(quals);
-	foreach(varClauseCell, varClauseList)
+	if (relationRestriction == NULL)
 	{
-		Var *column = (Var *) lfirst(varClauseCell);
-		if (equal(column, partitionColumn))
+		Relation relation = heap_open(newRte->relid, AccessShareLock);
+		TupleDesc tupleDescriptor = RelationGetDescr(relation);
+		int columnCount = tupleDescriptor->natts;
+		int columnIndex = 0;
+		List *valuesList = NIL;
+		List *collationList = NIL;
+
+		newRte->rtekind = RTE_VALUES;
+
+		for (columnIndex = 0; columnIndex < columnCount; columnIndex++)
 		{
-			partitionColumnReferenceCount++;
+			FormData_pg_attribute *attributeForm = tupleDescriptor->attrs[columnIndex];
+			if (!attributeForm->attisdropped)
+			{
+				Const *constValue = makeNullConst(attributeForm->atttypid,
+												  attributeForm->atttypmod,
+												  attributeForm->attcollation);
+				valuesList = lappend(valuesList, constValue);
+				collationList = lappend_oid(collationList, attributeForm->attcollation);
+			}
 		}
+
+
+		newRte->values_lists = NIL;
+		newRte->values_lists = lappend(newRte->values_lists, valuesList);
+		newRte->values_collations = collationList;
+		newRte->alias = copyObject(newRte->eref);
+		heap_close(relation, AccessShareLock);
+
+		return false;
 	}
 
-	if (partitionColumnReferenceCount != 1)
+	Assert(relationRestriction != NULL);
+
+	shardIntervalList = relationRestriction->prunedShardIntervalList;
+
+	Assert(list_length(shardIntervalList) == 1);
+	shardInterval = (ShardInterval *) linitial(shardIntervalList);
+
+	shardId = shardInterval->shardId;
+	relationId = shardInterval->relationId;
+	relationName = get_rel_name(relationId);
+	AppendShardIdToName(&relationName, shardId);
+
+	schemaId = get_rel_namespace(relationId);
+	schemaName = get_namespace_name(schemaId);
+
+	/* XXX: why is that a good idea? */
+	if (strncmp(schemaName, "public", NAMEDATALEN) == 0)
 	{
-		return false;
+		schemaName = NULL;
 	}
 
+	ModifyRangeTblExtraData(newRte, CITUS_RTE_SHARD, schemaName,
+							relationName,
+							NIL);
+
+	return false;
+}
+
+
+/*
+ * RouterQueryJob creates a Job for the specified query to execute the
+ * provided single shard select task.
+ */
+static Job *
+RouterQueryJob(Query *query, Task *task, List *placementList)
+{
+	Job *job = NULL;
+	List *taskList = NIL;
+	TaskType taskType = task->taskType;
+
 	/*
-	 * We need to make sure there is at least one shard for this hash partitioned
-	 * query to be router plannable. We can not prepare a router plan if there
-	 * are no shards.
+	 * We send modify task to the first replica, otherwise we choose the target shard
+	 * according to task assignment policy. Placement list for select queries are
+	 * provided as function parameter.
 	 */
-	shardCount = ShardIntervalCount(distributedTableId);
-	if (shardCount == 0)
+	if (taskType == MODIFY_TASK)
```

Comment:
probably we should deal with this as a seperate PR when we revisit modify task processing after prepared statements work is merged.